diff --git a/src/main.rs b/src/main.rs
index 1cdf9cf..3859ac0 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -63,7 +63,7 @@ fn orbital(daemon: Daemon) -> Result<(), String> {
         .enable();
 
     let mut args = env::args().skip(1);
-    let display_path = args.next().ok_or("no display argument")?;
+    let display_path = env::var("DISPLAY").unwrap();
     let login_cmd = args.next().ok_or("no login manager argument")?;
 
     core::fix_env(&display_path).map_err(|_| "error setting env vars")?;
